### DISCUSSION TITLE
Possibility to only have certain fields included in relationship

### DISCUSSION THREAD

**Discussion by christian-reichart** [payloadcms/payload#1042] [Feature Requests & Ideas]
Use case:

I want to use the relationship field to create a href link to a page.

My collection holds a field called `content` that holds complex nested blocks. For the link I only need the `slug` field of the collection.

The idea would be to have the option do define a `fields` array for the relationship field that defines what fields are included in the related document(s).

**Comment by DanRibbens** [comment]
Quite simply you want to specify a projection to use on a relation. That makes a lot of sense!

Payload currently only deals with this concept with `depth`. This is completely binary; depth of 0 only gives the relationship id and anything more, returns the full related object.

I think you could achieve this in an `afterRead` hook on the relationship field so as to modify the returned relationship with only the fields you needed. If you do this I'd recommend always returning the field you've used in the `useAsTitle` of the related collection along with the `id`, and `slug` or your admin UI will not work. Also doing this would break GraphQL unless you conditionally handle those requests differently by the `req.payloadAPI` coming in.

We probably do want Payload to do what you've described in config and we would accept a PR, but this would require a pretty big change and will need some planning to not break functionality around querying and hooks.

Some open questions I have are:
Would you still be able to query collections on their relationship fields even if the array does not include that field?
Does this only apply to REST and Local APIs or would GraphQL schemas need to be defined for each relationship that uses this feature?

**Reply by DanRibbens** [reply]
I should also mention, you could use GraphQL to grab only the data you need too. Do you have a hard requirement around using REST?

**Reply by christian-reichart** [reply]
@DanRibbens  thanks for your reply. We specifically opted against GraphQL, because for our use case GraphQL would just mean more work most of the time. I guess for the problem I have it would make sense. Thanks for the idea with the 'afterRead' hook, I will try  this out.
I'll keep this mind however and discuss with my colleagues if we are able to do a PR some time in the future.

**Reply by DanRibbens** [reply]
Sounds good! 

Thinking about this more, it doesn't make sense to add this into Payload since GraphQL already handles this so much better.

**Reply by christian-reichart** [reply]
@DanRibbens  Quick question, since I'm now struggling to achieve this goal with the afterRead hook. Inside the hook the value is always just the relationship id as a string. If I try to return anything complex, it get's reduced to '[object Object]'. How would I go about changing the returned data inside the hook?

If I understand the code correctly, the return value is only populated *after* the afterRead hook.

https://github.com/payloadcms/payload/blob/686085496a5bb677361e72c5e6504ab232d9da66/src/fields/hooks/afterRead/relationshipPopulationPromise.ts#L16

https://github.com/payloadcms/payload/blob/686085496a5bb677361e72c5e6504ab232d9da66/src/fields/hooks/afterRead/relationshipPopulationPromise.ts#L37

**Reply by christian-reichart** [reply]
Also found out that for internal links inside the RichText field the problem with the full population is also present, so it is not only a custom issue I have for my usecase, but a more general issue. Maybe this adds some relevance to still consider a solution for us REST users. :D
